### PR TITLE
OpenAPI Python Client

### PR DESCRIPTION
**Pull request by nopcoder**
Part of #1725 
Changes made to the project Makefile to generate the python client code.
Added:
- `clients/python/lakefs/client.py` to enable a single client with access to all APIs
- `clients/python/build-package.sh` to build pypi package

### CODE REVIEW

File: .github/workflows/goreleaser.yaml [hunk 1]
```diff
@@ -32,4 +32,4 @@ jobs:
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
-          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
```
Comment (itaiad200) [inline .github/workflows/goreleaser.yaml:35]: Add newline.

File: .github/workflows/pypi.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    types: [published]
+
+jobs:
+  pypi:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: '1'
```
Comment (itaiad200) [inline .github/workflows/pypi.yaml:14]: This is the default behaviour.

File: .github/workflows/pypi.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      - name: Extract version
+        shell: bash
+        run: echo "::set-output name=tag::$(echo ${GITHUB_REF##*/} | sed s/^v//g)"
+        id: version
+      - name: Python re-generate client
+        run: make client-python PACKAGE_VERSION=${{ steps.version.outputs.tag }}
+      - name: Python build package
+        run: make package-python 
+      - name: Python publish package
+        if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
```
Comment (itaiad200) [inline .github/workflows/pypi.yaml:24]: How can it be a `push` event when you're only triggered by `published`?

File: .goreleaser.yml [hunk 4]
```diff
@@ -2,7 +2,7 @@ project_name: lakeFS
 before:
   hooks:
     - go mod download
-    - make gen
+    - make gen PACKAGE_VERSION={{ .Version }}
```
Comment (itaiad200) [inline .goreleaser.yml:5]: You use `${{ steps.version.outputs.tag }}` in `pypi.yaml`

File: Makefile [hunk 5]
```diff
@@ -7,6 +7,15 @@ NPM=$(or $(shell which npm), $(error "Missing dependency - no npm in PATH"))
 PROTOC_IMAGE="treeverse/protoc:3.14.0"
 PROTOC=$(DOCKER) run --rm -v $(shell pwd):/mnt $(PROTOC_IMAGE)
 
+# https://openapi-generator.tech
+OPENAPI_GENERATOR_IMAGE=openapitools/openapi-generator-cli:v5.1.0
+OPENAPI_GENERATOR=$(DOCKER) run --rm -v $(shell pwd):/mnt $(OPENAPI_GENERATOR_IMAGE)
+ifndef PACKAGE_VERSION
+	PACKAGE_VERSION=0.1.0.dev
```
Comment (itaiad200) [inline Makefile:14]: Use git's SHA instead?

File: Makefile [hunk 6]
```diff
@@ -91,6 +100,21 @@ go-install: go-mod-download ## Install dependencies
...
+	$(OPENAPI_GENERATOR) generate \
+		-i /mnt/$< \
+		-g python \
+		--package-name lakefs \
+		--additional-properties=infoName=Treeverse,infoEmail=services@treeverse.io,packageName=lakefs,packageVersion=$(PACKAGE_VERSION),projectName=lakefs,packageUrl=https://github.com/treeverse/lakeFS/tree/feature/python-client/clients/python \
+		-o /mnt/clients/python
+
+clients: client-python
+
+package-python:
```
Comment (itaiad200) [inline Makefile:113]: Shouldn't it depend on `client-python`?

File: clients/python/.gitignore [hunk 7]
```diff
@@ -0,0 +1,66 @@
+# Byte-compiled / optimized / DLL files
```
Comment (itaiad200) [inline clients/python/.gitignore:1]: Do we really need everything in this file?

File: clients/python/.gitlab-ci.yml [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# ref: https://docs.gitlab.com/ee/ci/README.html
```
Comment (itaiad200) [inline clients/python/.gitlab-ci.yml:1]: Can you add a comment about this file?
I have 0 understanding of why we need it.

File: .github/workflows/pypi.yaml (hunk 3 above)
Comment (nopcoder) [reply .github/workflows/pypi.yaml:24]: true, I was triggering it manually - I'll remove the first part and just keep the tags check

File: .goreleaser.yml (hunk 4 above)
Comment (nopcoder) [reply .goreleaser.yml:5]: Right, this was part of a test to use gorelease to publish the package - will remove it

File: Makefile (hunk 5 above)
Comment (nopcoder) [reply Makefile:14]: Needed a constat. The build process runs `gen` and `checks-validator` which generate the client and check if there is a diff the client's API folder for changes.
This is to make sure in the PR step that you build the python client after updating the swagger API.

File: clients/python/.gitignore (hunk 7 above)
Comment (nopcoder) [reply clients/python/.gitignore:1]: it is auto-generated by the openapi-generator

File: clients/python/.gitlab-ci.yml (hunk 8 above)
Comment (nopcoder) [reply clients/python/.gitlab-ci.yml:1]: I'll remove it and add it to the ignore list so the generator will not re-create it.

**Review by itaiad200 — APPROVED** [review]
Great work!

**Pull request merged**